cmd/podman/images/scp.go

```diff
@@ -0,0 +1,383 @@
+		}
+	} else { // else we default to remote or whatever the default connection is
+		connStr := cfg.Engine.ActiveService
+		if len(cfg.Engine.ServiceDestinations) == 0 {
+			logrus.Warnf("Unknown connection name given. Please use system connection add to specify the default remote socket location")
+		}
+
+		for name, data := range cfg.Engine.ServiceDestinations {
+			if name == connStr {
+				iden = data.Identity
+				url = data.URI
+			}
+		}
+		dst := config.Destination{}
+		if iden != "" {
+			dst.URI = url
+			dst.Identity = iden
+		} else {
+			dst.URI = url
+		}
+		serv[connStr] = dst
+	}
+	optionsTunnel := entities.PodmanConfig{}
+	confR.Engine = config.EngineConfig{Remote: true, CgroupManager: "cgroupfs", ServiceDestinations: serv} // pass the service dest (either remote or something else) to engine
+	if iden != "" {
+		optionsTunnel = entities.PodmanConfig{Config: confR, EngineMode: entities.TunnelMode, Remote: true, Identity: iden, URI: url}
+	} else {
+		optionsTunnel = entities.PodmanConfig{Config: confR, EngineMode: entities.TunnelMode, Remote: true, URI: url}
+	}
+	if scpOpts.ToFrom {
+		abiErr := abiEng.Save(context.Background(), args[0], tags, scpOpts.Save) // save the image locally before loading it on remote, local, or ssh
+		if abiErr != nil {
+			errors.Wrapf(abiErr, "could not save image as specified")
+		}
+	}
+	uri, err := urlP.Parse(url) // create an actual url to pass to exec command
+	if err != nil {
+		return err
+	}
+	if uri.User.Username() == "" {
+		if uri.User, err = getUserInfo(uri); err != nil {
+			return err
+		}
+	}
+	rem := cmd.Root().Flags().Lookup("remote").Value // if remote was specified globally...
+	remS := rem.String()
+	if remS == "true" {
+		tunEng, err := infra.NewImageEngine(&optionsTunnel) // means we are doing a tunnel load
+		if err != nil {
+			return (errors.Wrapf(err, "could not make tunnel engine"))
+		}
+		tunOut, err := tunEng.Load(context.Background(), scpOpts.Load)
+		if err != nil {
+			errors.Wrapf(err, "could not load saved image")
+		}
+		if tunOut != nil {
+			out := tunOut.Names
+			if len(tunOut.Names) == 0 {
+				return errors.Wrapf(errors.New("Invalid Image given"), "could not get image")
+			}
+			fmt.Println("Loaded image(s): " + strings.Join(out, ","))
+		}
+	} else if len(args) == 2 && scpOpts.ToFrom { // else if a remote ssh connection was specified, means we need to go to exec command
+		rep, err := execCommand(iden, uri, scpOpts.Save.Output, scpOpts.Load.Input, args[0], scpOpts.ToFrom)
+		if err != nil {
+			return err
+		}
+		fmt.Println(rep)
+	} else if !scpOpts.ToFrom { // else if we want to load FROM the remote
+		_, err := execCommand(iden, uri, scpOpts.Save.Output, scpOpts.ImageName, args[0], scpOpts.ToFrom)
+		if err != nil {
+			return err
+		}
+		if len(args) == 2 && !(strings.Contains(args[1], "::") && strings.Contains(args[0], "::")) { // if we don't want to do a remote -> remote then simply load
```

Comment:
Way down here is also too deep to do argument parsing of this kind, especially when the arguments are not even used anywhere inside the conditional, and only exist as an “action at a distance” related to some other earlier condition. (The earlier one I commented on with “Also…”? Maybe?)

It would be very hard to maintain consistency of those two conditions.

---

Consider (this is not a recommendation, I didn’t read the code enough for that, just an avenue to investigate):
- A first “parse arguments” step that uses `scpOpts` and `args`, and turns them into local Go variables; after that “parse arguments” step, nothing ever refers to `args`, and instead uses those local variables. (Those variables might be booleans, or even an enum to determine the whole operation copy-to-remote / copy-from-remote / … ).
- A similar “parse arguments” step + individual functions that perform the different modes, with parsed options passed possibly as parameters.

(Also, at ~280 lines , the function is rather big. Along with how it does several different things, it might well make sense to split it into smaller units.)

Reply:
> to determine the whole operation copy-to-remote / copy-from-remote / …

to determine WHETHER the whole operation IS copy-to-remote / copy-from-remote / …